### PR TITLE
Dread Castles Gen, Fix, BlockSet and mobs, tripwires, ghosts, graveyard, cannoli and tons of bug fixes

### PR DESCRIPTION
**Pull request by S-Leader**

### CODE REVIEW

File: src/main/java/com/github/alexthe666/iceandfire/IceAndFireConfig.java [hunk 1]
```diff
@@ -174,6 +174,12 @@ public class IceAndFireConfig {
     public boolean explosiveDragonBreath = false;
     public float weezerTinkersDisarmChance = 0.2F;
     public boolean chunkLoadSummonCrystal = true;
+    public int ballistaBaseDamage = 10;
+    public double ghostMaxHealth;
+    public double ghostAttackStrength;
+    public boolean generateGraveyards = true;
+    public int generateGraveyardChance = 16;
+    public boolean ghostSpawnFromPlayerDeaths;
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/IceAndFireConfig.java]: Missing config options for some of these. Also, set default values for the ghost* values.

File: src/main/java/com/github/alexthe666/iceandfire/block/keletu/BlockGraveyardSoil.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        setRegistryName(IceAndFire.MODID, "graveyard_soil");
+    }
+
+
+    @Override
+    public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand) {
+        if (!worldIn.isRemote) {
+            if (!worldIn.isAreaLoaded(pos, 3))
+                return;
+            if (!worldIn.isDaytime() && !worldIn.getBlockState(pos.up()).causesSuffocation() && rand.nextInt(9) == 0 && worldIn.getDifficulty() != EnumDifficulty.PEACEFUL) {
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/block/keletu/BlockGraveyardSoil.java]: Potentially add a config option for ghost spawn chance

File: src/main/java/com/github/alexthe666/iceandfire/client/render/entity/RenderDreadKnightRoyal.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    protected ResourceLocation getEntityTexture(EntityDreadKnightRoyal entity) {
+        switch (entity.getArmorVariant()){
+            case 0:
+                return TEXTURE_0;
+            case 1:
+                return TEXTURE_1;
+            case 2:
+                return TEXTURE_2;
+                default:
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/client/render/entity/RenderDreadKnightRoyal.java]:
```suggestion
            default:
```

File: src/main/java/com/github/alexthe666/iceandfire/client/render/entity/RenderIceCharge.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+	public RenderIceCharge() {
+		super(Minecraft.getMinecraft().getRenderManager());
+	}
+
+	@Override
+	public void doRender(EntityBallistaArrow arrow, double a,
+			double b, double c, float d,
+			float e) {
+
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/client/render/entity/RenderIceCharge.java:18]: is this supposed to be empty?

File: src/main/java/com/github/alexthe666/iceandfire/entity/EntityDreadMob.java [hunk 5]
```diff
@@ -23,11 +18,18 @@
...
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import net.minecraft.world.storage.loot.LootTableList;
+
+import javax.annotation.Nullable;
+import java.util.UUID;
 
 public class EntityDreadMob extends EntityMob implements IDreadMob {
 
+	public static final ResourceLocation LOOT = LootTableList.register(new ResourceLocation("iceandfire", "dread_queen"));
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/entity/EntityDreadMob.java]: shouldn't this go into EntityDreadQueen class?

File: src/main/java/com/github/alexthe666/iceandfire/entity/EntityDreadQueen.java [hunk 6]
```diff
@@ -58,8 +51,13 @@ protected void initEntityAI() {
...
     }
 
+    @Nullable
+    protected ResourceLocation getLootTable() {
+        return LOOT;
+    }
+
     /**
-     * Pabilo & Carver, Quality Coding™
+     * Pabilo & Carver, Quality Coding(tm)
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/entity/EntityDreadQueen.java:64]: i like this change

File: src/main/java/com/github/alexthe666/iceandfire/entity/EntityGhost.java [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+                        double d5 = EntityGhost.this.getAttackTarget().posZ - EntityGhost.this.posZ;
+                        EntityGhost.this.rotationYaw = -((float) MathHelper.atan2(d4, d5)) * (180F / (float) Math.PI);
+                        EntityGhost.this.renderYawOffset = EntityGhost.this.rotationYaw;
+                    }
+                }
+            }
+        }
+
+        /*
+            public void tick() {
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/entity/EntityGhost.java:417]: What is the origin/purpose of this commented code?

File: src/main/java/com/github/alexthe666/iceandfire/entity/EntityGhostSwordTest.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
+package com.github.alexthe666.iceandfire.entity;
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/entity/EntityGhostSwordTest.java]: What is the purpose of this file?

File: src/main/java/com/github/alexthe666/iceandfire/entity/ai/DragonAIReturnToRoost.java [hunk 9]
```diff
@@ -12,14 +12,20 @@ public DragonAIReturnToRoost(EntityDragonBase entityIn) {
...
+        return this.dragon.canMove() && this.dragon.lookingForRoostAIFlag
+                && (dragon.getAttackTarget() == null || !dragon.getAttackTarget().isEntityAlive())
+                && dragon.getHomePosition() != null
+                && DragonUtils.isInHomeDimension(dragon)
+                && dragon.getDistanceSquared(IAFMath.copyCentered(dragon.getHomePosition())) > dragon.width * dragon.width;
+    }
+
     @Override
     public void updateTask() {
-        this.dragon.getHomePosition();
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/entity/ai/DragonAIReturnToRoost.java:22]: Could this be a cause of #48 ? (and thus would fix it?)

File: src/main/java/com/github/alexthe666/iceandfire/entity/tile/keletu/TileEntityDreadPiston.java [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+public class TileEntityDreadPiston extends TileEntityPiston {
+
+    public TileEntityDreadPiston()
+    {
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/entity/tile/keletu/TileEntityDreadPiston.java]: Why is there a default constructor here?

File: src/main/java/com/github/alexthe666/iceandfire/event/ServerEvents.java [hunk 11]
```diff
@@ -555,18 +595,12 @@ public void onLivingHurt(LivingHurtEvent event) {
     				}
     				break;
     			case 2:
-    				if(target.getCreatureAttribute() != EnumCreatureAttribute.ARTHROPOD || target instanceof EntityDeathWorm) {
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/event/ServerEvents.java]: Can you revert this change (and presumably the corresponding change in ItemRegistry) so it'll be easier to merge with my most recent change on the main branch? (Don't merge that commit in to this manually, though.)

File: src/main/java/com/github/alexthe666/iceandfire/event/WorldGenDreadDimension.java [hunk 12]
```diff
@@ -32,12 +29,11 @@ public void generate(Random random, int chunkX, int chunkZ, World world, IChunkG
             BlockPos height = getHeight(world, new BlockPos(x, 0, z));
 
             if (BiomeDictionary.hasType(world.getBiome(height), BiomeDictionary.Type.COLD) && BiomeDictionary.hasType(world.getBiome(height), BiomeDictionary.Type.SNOWY)) {
-                if (random.nextInt(1000) == 0) {
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/event/WorldGenDreadDimension.java:35]: What is the purpose for removing this? (I think I know why but I'm double-checking)

File: src/main/java/com/github/alexthe666/iceandfire/item/IafItemRegistry.java [hunk 13]
```diff
@@ -306,7 +307,7 @@ public class IafItemRegistry {
     @GameRegistry.ObjectHolder(IceAndFire.MODID + ":myrmex_desert_sword")
     public static Item myrmex_desert_sword = new ItemModSword(myrmexChitin, "myrmex_desert_sword", "iceandfire.myrmex_desert_sword", 1);
     @GameRegistry.ObjectHolder(IceAndFire.MODID + ":myrmex_desert_sword_venom")
-    public static Item myrmex_desert_sword_venom = new ItemModSword(myrmexChitin, "myrmex_desert_sword_venom", "iceandfire.myrmex_desert_sword_venom", 2);
+    public static Item myrmex_desert_sword_venom = new ItemModSword(myrmexChitin, "myrmex_desert_sword_venom", "iceandfire.myrmex_desert_sword_venom", 5);
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/item/IafItemRegistry.java]: As mentioned before- revert this change to make it easier to rebase onto main branch

File: src/main/java/com/github/alexthe666/iceandfire/item/ItemModSword.java [hunk 14]
```diff
@@ -44,21 +42,21 @@ public void addInformation(ItemStack stack, @Nullable World worldIn, List<String
...
     		break;
-    	case 4: 
+    	case 3:
     		ItemUtil.getIceDragonsteelComment(tooltip);
     		break;
-    	case 5:
+    	case 4:
     		ItemUtil.getLightningDragonsteelComment(tooltip);
     		break;
+		case 5:
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/item/ItemModSword.java]: As mentioned earlier, revert this to make rebase easier

File: src/main/java/com/github/alexthe666/iceandfire/world/gen/WorldGenCastle.java [hunk 15]
```diff
@@ -44,31 +46,44 @@ private static boolean canHeightSkipBlock(BlockPos pos, World world) {
     }
 
     public boolean generate(World worldIn, Random rand, BlockPos position) {
+
+        if(!checkIfCanGenAt(worldIn, position, 1000, 1000))
+            return false;
+
         position = position.add(rand.nextInt(8) - 4, 1, rand.nextInt(8) - 4);
-        int totalW = 64, totalH = 64, totalD = 64; // TODO: 27.06.2022 change
+        int totalW = 274, totalH = 130, totalD = 274; // TODO: 27.06.2022 change
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/world/gen/WorldGenCastle.java]: Is this TODO still applicable?

File: src/main/java/com/github/alexthe666/iceandfire/world/gen/WorldGenCastle.java [hunk 16]
```diff
@@ -44,31 +46,44 @@ private static boolean canHeightSkipBlock(BlockPos pos, World world) {
     }
 
     public boolean generate(World worldIn, Random rand, BlockPos position) {
+
+        if(!checkIfCanGenAt(worldIn, position, 1000, 1000))
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/world/gen/WorldGenCastle.java:51]: Add a code comment above this line explaining what this does

File: src/main/java/com/github/alexthe666/iceandfire/world/gen/WorldGenCastle.java [hunk 17]
```diff
@@ -44,31 +46,44 @@ private static boolean canHeightSkipBlock(BlockPos pos, World world) {
...
+                        for(int u = 0; u < totalW; u++)
+                            worldIn.setBlockState(origin.add(t, -1, u), IafBlockRegistry.dread_stone.getDefaultState(), 2);
+
+                    template.addBlocksToWorld(worldIn, pos, new DreadCastleProcessor(), new PlacementSettings(), 2);
                 }
 
         return true;
     }
 
     // TODO: 27.06.2022 implement a proper check
```
Comment (democat3457) [inline src/main/java/com/github/alexthe666/iceandfire/world/gen/WorldGenCastle.java]: Also, is this TODO still applicable?

File: src/main/resources/assets/iceandfire/lang/en_us.lang [hunk 18]
```diff
@@ -448,12 +466,16 @@ entity.dread_scuttler.name=Dread Scuttler
 entity.dread_lich.name=Dread Lich
 entity.dread_lich_skull.name=Dread Lich Skull
 entity.dread_knight.name=Dread Knight
+entity.dread_knight_royal.name=Dread Knight Royal
```
Comment (democat3457) [inline src/main/resources/assets/iceandfire/lang/en_us.lang]:
```suggestion
entity.dread_knight_royal.name=Royal Dread Knight
```

File: src/main/resources/assets/iceandfire/lang/en_us.lang [hunk 19]
```diff
@@ -448,12 +466,16 @@ entity.dread_scuttler.name=Dread Scuttler
...
+entity.ballista.name=Castle Ballista
 entity.dread_horse.name=Dread Knight Horse
 entity.black_frost_dragon.name=Black Frost
 entity.dread_queen.name=Dread Queen
 entity.if_hydra.name=Hydra
 entity.hydra_breath.name=Hydra Breath
 entity.hydra_arrow.name=Hydra Arrow
+entity.ghost.name=Ghost
+entity.ghost_sword.name=Spirit Sword
```
Comment (democat3457) [inline src/main/resources/assets/iceandfire/lang/en_us.lang]: Should this be Phantasmal Blade?

File: src/main/resources/assets/iceandfire/lang/en_us.lang [hunk 20]
```diff
@@ -606,7 +628,9 @@ message.iceandfire.dragonSitName= is now Staying.
 message.iceandfire.dragonSleepName= is now Sleeping.
 message.iceandfire.dragonGrown=This dragon has grown to Stage 
 message.iceandfire.dragonGrownEnd=!
-message.iceandfire.dragonGrownName= has grown to Stage 
+message.iceandfire.dragonGrownName= has grown to Stage
+message.iceandfire.toocold=Seems i can't eat this at this dimension
+message.iceandfire.cantfly=The storm forced you down
```
Comment (democat3457) [inline src/main/resources/assets/iceandfire/lang/en_us.lang]:
```suggestion
message.iceandfire.toocold=Seems I can't eat this in this dimension.
message.iceandfire.cantfly=The storm forced you down.
```

**Review by democat3457 — CHANGES_REQUESTED** [review]
Overall looks great, but there's some minor changes that should be addressed first. Also, a few comments checking some things.

File: src/main/java/com/github/alexthe666/iceandfire/client/render/entity/RenderIceCharge.java (hunk 4 above)
Comment (S-Leader) [reply src/main/java/com/github/alexthe666/iceandfire/client/render/entity/RenderIceCharge.java:18]: hmm, yes

File: src/main/java/com/github/alexthe666/iceandfire/event/WorldGenDreadDimension.java (hunk 12 above)
Comment (S-Leader) [reply src/main/java/com/github/alexthe666/iceandfire/event/WorldGenDreadDimension.java:35]: because now castle will spawn each 2000 block instead of spawn with randomly

File: src/main/java/com/github/alexthe666/iceandfire/entity/EntityDreadQueen.java (hunk 6 above)
Comment (S-Leader) [reply src/main/java/com/github/alexthe666/iceandfire/entity/EntityDreadQueen.java:64]: me 2 :)

**Review by democat3457 — APPROVED** [review]
